File: samples/cpp/text_generation/greedy_causal_lm.cpp
// Copyright (C) 2023-2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include "openvino/genai/llm_pipeline.hpp"
#include "filesystem"
int main(int argc, char* argv[]) try {
    std::cout << "Init\n";
    if (2 > argc)
        //throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> \"<PROMPT>\" <TOKENIZER_PATH>");
        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> \"<PROMPT>\"");
    
    std::string models_path = argv[1];
    std::string prompt = argv[2];
    //std::string tokenizer_path = argv[3];
    std::string device = "CPU";  // GPU can be used as well
    //ov::genai::Tokenizer tokenizer(tokenizer_path);
    ov::AnyMap properties;
    properties["ATTENTION_BACKEND"] = "SDPA";
    //std::filesystem::path model_path(models_path);
    /*
    if (std::filesystem::exists(model_path.parent_path() / "openvino_model.xml")) {
        models_path = model_path.parent_path();
    } else {
        models_path = model_path;
    }
    properties[ov::cache_dir.name()] = "model_cache";
    properties[ov::genai::enable_save_ov_model.name()] = true;
    
    ov::genai::LLMPipeline pipe(models_path, tokenizer, device, properties);
    */
    std::cout << "Create pipeline start\n";
    ov::genai::LLMPipeline pipe(models_path, device, properties);
    std::cout << "Create pipeline end\n";
    ov::genai::GenerationConfig config;
    config.max_new_tokens = 100;
    std::cout << "Generating text with prompt: " << prompt << std::endl;
    std::string result = pipe.generate(prompt, config);
    std::cout << result << std::endl;
} catch (const std::exception& error) {
    try {
        std::cerr << error.what() << '\n';
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
} catch (...) {
    try {
        std::cerr << "Non-exception object thrown\n";
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
}

File: samples/python/text_generation/greedy_causal_lm.py
#!/usr/bin/env python3
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import openvino_genai


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('model_dir')
    parser.add_argument('prompt')
    args = parser.parse_args()

    device = 'CPU'  # GPU can be used as well
    # pipe = openvino_genai.LLMPipeline(args.model_dir, device)
    scheduler_confg = openvino_genai.SchedulerConfig()
    generation_config = openvino_genai.GenerationConfig()
    generation_config.max_new_tokens = 1000
    props = {}
    # ... adjust scheduler and geneartion configs as necessary ..

    scheduler_confg.use_cache_eviction = False  # enables_cache_evction

    # optional - configure eviction parameters
    scheduler_confg.cache_eviction_config = openvino_genai.CacheEvictionConfig(
        start_size=32,
        recent_size=128,
        max_cache_size=672,
        aggregation_mode=openvino_genai.AggregationMode.NORM_SUM)

    print("Init pipeline ...\n")
    pipe = openvino_genai.ContinuousBatchingPipeline(args.model_dir,
                                                     scheduler_confg,
                                                     device,
                                                     {},
                                                     props
                                                     )
    # config = openvino_genai.GenerationConfig()
    # config.max_new_tokens = 100

    # print(pipe.generate(args.prompt, config))
    prompt_list = [args.prompt] * 5
    print("Start generation ...\n")
    result = pipe.generate(prompt_list, [generation_config] * len(prompt_list))


if '__main__' == __name__:
    main()
